src/libsyntax/ext/expand.rs

```diff
@@ -876,14 +893,50 @@ pub fn expand_type(t: P<ast::Ty>, fld: &mut MacroExpander) -> P<ast::Ty> {
     fold::noop_fold_ty(t, fld)
 }
 
+macro_rules! macro_expander {
+    ($(($expand:ident, $ty: ty)),*) => {
+        pub struct MacroExpander<'a, 'b:'a> {
+            pub cx: &'a mut ExtCtxt<'b>,
+
+            //Expander functions
+            $(pub $expand: Rc<Box<Fn(&mut MacroExpander, $ty) -> $ty>>,)*
+        }
+    }
+}
+
 /// A tree-folder that performs macro expansion
-pub struct MacroExpander<'a, 'b:'a> {
-    pub cx: &'a mut ExtCtxt<'b>,
+macro_expander! {
+    (expand_pat, P<ast::Pat>),
+    (expand_type, P<ast::Ty>),
+    (expand_expr, P<ast::Expr>),
+    (expand_stmt, SmallVector<ast::Stmt>),
+    (expand_item, SmallVector<P<ast::Item>>),
+    (expand_impl_item, SmallVector<ast::ImplItem>),
+    (expand_trait_item, SmallVector<ast::TraitItem>),
+    (expand_opt_expr, Option<P<ast::Expr>>)
+}
```

Comment:
I might be missing something, but it seems these macros don't save us much - seems like all we save is having to write the full signature for the expander function, which would be better served by a generic type alias.